### PR TITLE
Macroblock generator

### PR DESCRIPTION
**Pull request by junmendoza**
### Purpose
This implements the macroblock generator and macroblock generator test cases 
The generator is integrated into the codegen.
The generator takes in a snapshot of the program and uses it to generate the macroblocks

#### References

Macroblock docs
https://docs.google.com/document/d/1AgpoMH-lG3XMVsS42UBXRCLcdkXZVC3TrDsWxMzInBw
https://docs.google.com/document/d/1KV5MvKBJG2rr82f4MkpsBeLhNcKo5M-FAc0L505BN-I

#### Testing strategy

1. MacroblockGenerator tests are added - MacroblockGeneratorTests.cs
2. All existing tests pass. This means that macroblocks are being generated for all existing testcases

### Review staging

- [x] Architectural Design Review

### Reviewers

PTAL @lukechurch @ke-yu 

Note that ive temporarily removed the Sequencer and all code relevant to macroblock execution.
Ill add this in the next PR

### CODE REVIEW

File: src/Engine/ProtoAssociative/CodeGen.cs [hunk 1]
```diff
@@ -4233,10 +4232,15 @@ public override int Emit(ProtoCore.AST.Node codeBlockNode, ProtoCore.Associative
...
-                codeBlock.instrStream.dependencyGraph.GetGraphNodesAtScope(Constants.kInvalidIndex, Constants.kGlobalScope));
-        
+
+            List<GraphNode> nodesInScope = codeBlock.instrStream.dependencyGraph.GetGraphNodesAtScope(Constants.kInvalidIndex, Constants.kGlobalScope);
+            ProtoCore.AssociativeEngine.Utils.BuildGraphNodeDependencies(nodesInScope);
+
+            // Generate the macroblock fragments at the global scope
+            if (codeBlock.codeBlockId == 0)
+            {
+                GenerateAndCacheMacroblocks(nodesInScope);
```
Comment (junmendoza) [inline src/Engine/ProtoAssociative/CodeGen.cs]: The snapshot of the program is the global scope

File: src/Engine/ProtoCore/AssociativeGraph.cs [hunk 2]
```diff
@@ -384,7 +384,7 @@ public static List<AssociativeGraph.GraphNode> GetGraphNodesFromAST(ProtoCore.DS
         public static bool IsGlobalScopeDirty(Executable exe)
         {
             Validity.Assert(exe != null);
-            var graph = exe.instrStreamList[0].dependencyGraph;
+            var graph = exe.GetInstructionStream(0).dependencyGraph;
```
Comment (junmendoza) [inline src/Engine/ProtoCore/AssociativeGraph.cs]: GetInstructionStream is minor cleanup work to isolate references to the instruction stream.

File: src/Engine/ProtoCore/DSASM/Executive.cs [hunk 3]
```diff
@@ -2692,8 +2707,7 @@ private void ExecuteDebug(int exeblock, int entry, List<Instruction> breakpoints
             }
         }
 
-
-        private void Execute(int exeblock, int entry, Language language = Language.kInvalid)
+        private void ExecuteInternal(int exeblock, int entry, Language language = Language.kInvalid)
```
Comment (junmendoza) [inline src/Engine/ProtoCore/DSASM/Executive.cs]: Minor refactor - renaming Execute to ExecuteInternal. This signifies that these calls are only from within the current Executive

File: test/Engine/ProtoTest/Macroblock/MacroblockGeneratorTests.cs [hunk 4]
```diff
@@ -0,0 +1,162 @@
...
+        public void TestSequential01()
+        {
+            const string code =
+                @"
+    a = 1;
+";
+            CompileAndGenerateMacroblocks(code);
+
+            // Verify
+            VerifyMacroblockCount(1);
```
Comment (junmendoza) [inline test/Engine/ProtoTest/Macroblock/MacroblockGeneratorTests.cs]: Verifies the number of generated macroblocks fromt he compiled code

File: test/Engine/ProtoTest/Macroblock/MacroblockGeneratorTests.cs [hunk 5]
```diff
@@ -0,0 +1,162 @@
...
+            VerifyGraphNodesInMacroblock(0, 1);
```
Comment (junmendoza) [inline test/Engine/ProtoTest/Macroblock/MacroblockGeneratorTests.cs]: Verifies the number of graphnodes associated with the macroblock. Im open to suggestions for additional verification methods

File: test/Engine/ProtoTestConsoleRunner/Program - Copy.cs [hunk 6]
```diff
@@ -1,86 +0,0 @@
-﻿
-#define __RUN_TESTFILE
```
Comment (junmendoza) [inline test/Engine/ProtoTestConsoleRunner/Program - Copy.cs]: Removing this file which as accidentally checked in

**Comment by ikeough** [comment]
@lukechurch @ke-yu This has been open a long time. Is it relevant? If so, who is going to finish it? If not, can we close it?

**Comment by ke-yu** [comment]
@ikeough , @lukechurch is reviewing it. 

@lukechurch , please let me know what do you think. If you've finished the review, i will merge it.

**Pull request closed without merging**